File: producer_consumer_python_threads/producer_consumer_threads.py
# Classic producer-consumer problem solution using condition variables.
# one producer and ten consumers access a queue.
# Producers generate a total of 100 messages each,
# and the consumer consumes all of them.
# Random sleep time is used to make execution less deterministic.

import threading
import collections
from random import randint
from time import sleep

MAX_QUEUE_SIZE=50
NUMBER_OF_PRODUCERS=10
ITEMS_TO_CONSUME=1000

# Multi-thread Interface
class MTI:
    def __init__ (self, size):
        self.size = size
        # signal consumer
        self.cv_queue_is_not_empty = threading.Condition()
        # signal producer
        self.cv_queue_is_not_full = threading.Condition()
        # lock the queue (not really required)
        self.rwlock = threading.Lock()
        # note deque operations are thread safe
        self.thequeue = collections.deque([""], size)
        self.thequeue.clear()

    def consume(self):
        self.cv_queue_is_not_empty.acquire()
        while len(self.thequeue) == 0:
            self.cv_queue_is_not_empty.wait()
        # deque.pop is thread is thread safe, but let's assume it is not.
        # "To make pop thread safe" rwlock is used
        self.rwlock.acquire()
        res=self.thequeue.pop()
        self.rwlock.release()
        self.cv_queue_is_not_empty.release()
        self.cv_queue_is_not_full.acquire()
        self.cv_queue_is_not_full.notify()
        self.cv_queue_is_not_full.release()
        return res
        
    def produce(self, what, idd):
        # only one producer at a time can enter
        self.cv_queue_is_not_full.acquire()
        # wait for free space in the queue
        while self.size <= len(self.thequeue):
            self.cv_queue_is_not_full.wait()
        #print("PRODUCER release lock" + idd + "number of els " + str(self.count) + " qlen " + str(len(self.thequeue)))
        # deque.append is thread safe, but let's assume it is not...
        # acquire exclusive access to the queue;
        # privent the consumer from accessing the queue.
        self.rwlock.acquire()
        self.thequeue.append(what)
        # The item has been produced, allow the consumer to access
        self.rwlock.release()
        # allow the other producers to enter
        self.cv_queue_is_not_full.release()
        # notify the consumer
        self.cv_queue_is_not_empty.acquire()
        self.cv_queue_is_not_empty.notify()
        self.cv_queue_is_not_empty.release()


def producer (pr_id, mti):
    ctr=1
    while ctr<=ITEMS_TO_CONSUME/NUMBER_OF_PRODUCERS:
         mti.produce("abc" + str((pr_id + 1) * 100000 + ctr), str(pr_id))
         ctr=ctr+1
    return

def consumer (mti):
    ctr=0
    all_msg=[]
    while ctr<ITEMS_TO_CONSUME:
        msg=mti.consume()
        all_msg.append(msg)
        sleep(randint(0,100)/10000)
        ctr+=1
    print(sorted(all_msg))
    return

if __name__ == "__main__":
    mti = MTI(MAX_QUEUE_SIZE)
    consumer_thread = threading.Thread(target=consumer, args=(mti,))
    producers = []
    i = 0
    while i < NUMBER_OF_PRODUCERS:
        producers.append(threading.Thread(target=producer, args=(i, mti)))
        i+=1

    consumer_thread.start()
    for prod in producers:
        prod.start();
    consumer_thread.join();
    for prod in producers:
        prod.join();

File: producer_consumer_cpp_threads/producer_consumer_threads.cpp
// classic producer-consumer problem solution using C++11 condition variables.
// The producer-consumer queue size is limited.
// One producer and ten consumers access a queue.
// Each producer generates a total of10 messages,
// and the consumer consumes all of them.
// Random sleep time is used to make execution less deterministic.
// Build instructions:
// g++ -o producer_consumer_threads -std=c++11 producer_consumer_threads.cpp

#include <cassert>
#include <chrono>
#include <condition_variable>
#include <cstdint>
#include <iostream>
#include <mutex>
#include <queue>
#include <random>
#include <thread>

using namespace std;

static const uint32_t MAX_QUEUE_SIZE = 5;
static const uint32_t MAX_MESSAGE_SIZE = 100;
static const uint32_t NUMBER_OF_PRODUCERS = 10;

static uint32_t items_consumed = 0;
static uint32_t items_produced = 0;

typedef struct {char buffer[MAX_MESSAGE_SIZE];} QUEUE_EL_T;

queue<QUEUE_EL_T> pcqueue;
mutex rwlock;
condition_variable cv_queue_not_full;
condition_variable cv_queue_not_empty;
mutex mtx_queue_not_full;
mutex mtx_queue_not_empty;
// correctness verification
static uint32_t number_of_elements = 0;

// Let each message has this header.
// Nothing else shall be sent/received but the header.
struct MSG {
    int pid;
    int mid;
};

void producer (int pr_id)
{
    random_device rd;
    mt19937 mt(rd()); // random engige
    uniform_int_distribution<int> dist(0, 10); // distribution
    for (int i = 0; i < 10; i++) {
        // until the queue is full wait for the consumer to free queue elements
        unique_lock<mutex> lock_queue_not_full(mtx_queue_not_full);
        while (pcqueue.size() >= MAX_QUEUE_SIZE) {
            cv_queue_not_full.wait(lock_queue_not_full);
            // sleep 0 - 10 ms
            this_thread::sleep_for (chrono::milliseconds(dist(mt)));
        }
        // get the lock for the queue
        // and produce an item
        rwlock.lock();
        QUEUE_EL_T el;
        MSG *pmsg = (MSG *) &el;
        pmsg->pid = pr_id;
        pmsg->mid = i;
        pcqueue.push(el);
        items_produced++;
        number_of_elements++;
        // check the invariant
        assert(number_of_elements <= MAX_QUEUE_SIZE);
        // Queue updates are done. Other threads can access the queue. Release the lock.
        rwlock.unlock();
        // In case the consumer was waiting on empty queue,
        // notify the consumer because the queue is not empty anymore
        cv_queue_not_empty.notify_all();
    }
}

void consumer ()
{
    random_device rd;
    mt19937 mt(rd());
    uniform_int_distribution<int> dist(0, 10);
    for (int i = 0; i < 100; i++) {
        unique_lock<mutex> lock_queue_not_empty(mtx_queue_not_empty);
        // wait for producers to produce
        while (pcqueue.size() == 0) {
            cv_queue_not_empty.wait(lock_queue_not_empty);
            this_thread::sleep_for (chrono::milliseconds(dist(mt)));
        }
        // read what was produced
        rwlock.lock();
        // read just one "message". It is possible to read all of them, of course.
        QUEUE_EL_T el = pcqueue.front();
        MSG *pmsg = (MSG *) &el;
        cout << "consumed: prod_id " << pmsg->pid << " mid " << pmsg->mid << " queue size " << pcqueue.size() << endl;
        pcqueue.pop();
        items_consumed++;
        number_of_elements--;
        assert(number_of_elements <= MAX_QUEUE_SIZE);
        rwlock.unlock();
        // notify producers if they were waiting on the full queue
        cv_queue_not_full.notify_all();
    }
    cout << "CONSUMER HAS RETIRED" << endl;   
}

main ()
{
    thread consumer_thread(consumer);
    thread *producers[NUMBER_OF_PRODUCERS];
    for (int i = 0; i < NUMBER_OF_PRODUCERS; i++) {
        producers[i] = new thread (producer, i);
    }
    consumer_thread.join();
    for (auto prod : producers)
        prod->join();
    assert(items_consumed == items_produced);
}
